[JENKINS-35459] - Prevent Duplicated elements with incorrect URL when using the search on DashBoard View

alexanderrtaylor opened the pull request:
Fixing the search bar not working for folders because Felix’s PR is
good: https://github.com/jenkinsci/jenkins/pull/2401 but needed to fix
some tests

https://issues.jenkins-ci.org/browse/JENKINS-35459

### Changelog entry

* JENKINS-35459, Bug: Prevent Duplicated elements with incorrect URL when using the search on DashBoard View

Commit:
Fixing JENKINS-35459 from PR-2401 and correcting tests

Fixing the search bar not working for folders because Felix’s PR is
good: https://github.com/jenkinsci/jenkins/pull/2401 but needed to fix
some tests

oleg-nenashev commented on core/src/main/java/hudson/model/ListView.java:
```diff
@@ -168,15 +170,24 @@ public boolean hasJobFilterExtensions() {
+
     /**
-     * Returns a read-only view of all {@link Job}s in this view.
+    * Returns a read-only view of all {@link Job}s in this view.
```
Breaks formatting 🐛

oleg-nenashev commented on core/src/main/java/hudson/model/ListView.java:
```diff
@@ -168,15 +170,24 @@ public boolean hasJobFilterExtensions() {
      * <p>
-     * This method returns a separate copy each time to avoid
-     * concurrent modification issue.
+     *     This method returns a separate copy each time to avoid
```
breaks correct formatting

oleg-nenashev commented on core/src/main/java/hudson/model/ListView.java:
```diff
@@ -168,15 +170,24 @@ public boolean hasJobFilterExtensions() {
-     * concurrent modification issue.
+     *     This method returns a separate copy each time to avoid
+     *     concurrent modification issue.
+     *     @Param recurse false not to recurse in ItemGroups
```
`@param` should be lowercase

oleg-nenashev commented on core/src/main/java/hudson/model/ListView.java:
```diff
@@ -216,6 +227,23 @@ public boolean hasJobFilterExtensions() {
+            }
+        });
+        // add the display name for each item in the search index
+        addDisplayNamesToSearchIndex(sib, getItems(false));
```
Why is there `false` here?

alexanderrtaylor replied:
Actually that is a good suggestion. I would think you want to recursively search over getItems. I am changing it to true and testing to see if that is behavior we want

oleg-nenashev reviewed (changes requested):
There are breaking formatting changes. If you are modifying the existing commits, you also need to keep the reference to the contributor in the commit history or to get his permission to not reference him. CC @fbelzunc

Commit:
Updated based on Oleg's suggestions

-Fixed formatting errors
-Made item search recursive

oleg-nenashev reviewed (approved):
🐝 , just minor comments mentioned above

aheritier commented:
LGTM 🐝

oleg-nenashev commented:
Needs feedback from @daniel-beck . Keeping on-hold

daniel-beck commented on test/src/test/java/hudson/search/SearchTest.java:
```diff
@@ -454,4 +454,44 @@ public void adminOnlyLinksNotShownToRegularUser() {
+        j.assertGoodStatus(result);
+
+        String content = result.getWebResponse().getContentAsString();
+        System.out.println(content);
```
leftover debug code?

oleg-nenashev replied:
Right, it should be Logger at least. Otherwise streams won't be syncronized

daniel-beck reviewed (commented):
Would be useful if comments (can be PR comments, not code comments) indicated how the tests would fail before this fix.

oleg-nenashev reviewed (changes requested):
🐛 as @daniel-beck noticed, logging in tests needs to be cleaned up

Commit:
Fixing based on suggestions from Oleg and Daniel

Fixing Javadoc info and removing the system.out.print line that was
used for debugging purposes

oleg-nenashev reviewed (commented):
@alexanderrtaylor Was there any specific reason to remove Javadoc of the public method?

alexanderrtaylor commented:
@oleg-nenashev 

You had commented: 
"oleg-nenashev 19 days ago Owner
Fine to remove custom Javadoc from what I see"

I thought that is what you were referring to but maybe I missed what you were saying?

oleg-nenashev reviewed (approved):
Or I just have no memory of this place...
Hence 🐝

oleg-nenashev commented:
@revuewbybees done

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

daniel-beck commented on core/src/main/java/hudson/model/ListView.java:
```diff
@@ -167,16 +169,24 @@ public boolean hasJobFilterExtensions() {
-    
+
+
+     public List<TopLevelItem> getItems() {
```
What is this for? Does this need `@Override`, `@Restricted` or `@since`?

daniel-beck replied:
Yep, lost `@Override` without explanation.

oleg-nenashev replied:
I will cleanup the Javadoc / Override changes after the merge

oleg-nenashev commented:
@daniel-beck do you require `@Override` to be restored?

oleg-nenashev commented:
It has 2 votes, so it's ready for merge.

daniel-beck commented:
@oleg-nenashev No strong opinion, but diff looked suspiciously like something newly added without restriction or doc.

daniel-beck commented:
What happened to

> If you are modifying the existing commits, you also need to keep the reference to the contributor in the commit history or to get his permission to not reference him. CC @fbelzunc

oleg-nenashev commented:
@alexanderrtaylor @fbelzunc ping regarding the contribution references. We will not merge the PR until there is a fix or approval

fbelzunc commented:
@oleg-nenashev @alexanderrtaylor I don't care about my previous commits :-) You can merge it!

oleg-nenashev commented:
@fbelzunc thanks for the response!
